src/transformers/models/superglue/modeling_superglue.py

```diff
@@ -0,0 +1,806 @@
+            list_matches_1,
+            list_matching_scores_0,
+            list_matching_scores_1,
+            pixel_values,
+            output_attentions=output_attentions,
+            output_hidden_states=output_hidden_states,
+        )
+
+        if not return_dict:
+            return tuple(
+                v
+                for v in [matches_mask, matches, matching_scores, keypoints, hidden_states, attentions]
+                if v is not None
+            )
+
+        return KeypointMatchingOutput(
+            loss=loss,
+            mask=matches_mask,
+            matches=matches,
+            matching_scores=matching_scores,
+            keypoints=keypoints,
+            hidden_states=hidden_states,
+            attentions=attentions,
+        )
+
+    def batch_output(
+        self,
+        batch_size,
+        list_attentions,
+        list_hidden_states,
+        list_keypoints_0,
+        list_keypoints_1,
+        list_matches_0,
+        list_matches_1,
+        list_matching_scores_0,
+        list_matching_scores_1,
+        pixel_values,
+        output_attentions,
+        output_hidden_states,
+    ):
+        maximum_matches = max(
+            [
+                max([matches_0.shape[1] for matches_0 in list_matches_0]),
+                max([matches_1.shape[1] for matches_1 in list_matches_1]),
+            ]
+        )
+        matches = torch.full(
+            (batch_size, 2, maximum_matches),
+            -1,
+            device=pixel_values.device,
+            dtype=torch.int,
+        )
+        matching_scores = torch.zeros((batch_size, 2, maximum_matches), device=pixel_values.device)
+        matches_mask = torch.zeros(
+            (batch_size, 2, maximum_matches),
+            device=pixel_values.device,
+            dtype=torch.int,
+        )
+        keypoints = torch.zeros(
+            (batch_size, 2, maximum_matches, 2),
+            device=pixel_values.device,
+        )
+        for i, (
+            _matches_0,
+            _matches_1,
+            _matching_scores_0,
+            _matching_scores_1,
+            _keypoints_0,
+            _keypoints_1,
+        ) in enumerate(
+            zip(
+                list_matches_0,
+                list_matches_1,
+                list_matching_scores_0,
+                list_matching_scores_1,
+                list_keypoints_0,
+                list_keypoints_1,
+            )
+        ):
+            matches[i, 0, : _matches_0.shape[1]] = _matches_0
+            matches[i, 1, : _matches_1.shape[1]] = _matches_1
+            matching_scores[i, 0, : _matching_scores_0.shape[1]] = _matching_scores_0
+            matching_scores[i, 1, : _matching_scores_1.shape[1]] = _matching_scores_1
+            matches_mask[i, 0, : _matches_0.shape[1]] = 1
+            matches_mask[i, 1, : _matches_1.shape[1]] = 1
+            keypoints[i, 0, : _keypoints_0.shape[1], :] = _keypoints_0
+            keypoints[i, 1, : _keypoints_1.shape[1], :] = _keypoints_1
+
```

Comment:
Regarding batching of the outputs, we don't want to batch together all of the attentions and hidden states into one big tensor. The pattern for other models is a tuple of tensors is returned, with each element in the tensor representing a layer or block of the model. Sorry if I wasn't clear about this earlier.

Reply:
Oh but it is not the case, we don't have a big tensor for all hidden_states and a big tensor for all attentions, it is still a tuple of hidden states and a tuple of attentions. What we are batching here is the multiple tuples of hidden states and attentions from the different image matching, together